File: PrimeSieveC/src/primes.c
#include <time.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <inttypes.h>

static bool* bits = NULL;
static size_t sieve_size;

static size_t count_primes(void) {
	size_t count = (sieve_size >= 2);
	for(size_t i = 3; i < sieve_size; i += 2) {
		if(bits[i]) {
			++count;
		}
	}
	return count;
}

static bool validate_results(void) {
	switch(sieve_size) {
	case 10:
		return count_primes() == 4;
	case 100:
		return count_primes() == 25;
	case 1000:
		return count_primes() == 168;
	case 10000:
		return count_primes() == 1229;
	case 100000:
		return count_primes() == 9592;
	case 1000000:
		return count_primes() == 78498;
	case 10000000:
		return count_primes() == 664579;
	case 100000000:
		return count_primes() == 5761455;
	case 1000000000:
		return count_primes() == 50847534;
	case 10000000000:
		return count_primes() == 455052511;
	default:
		return false;
	}
}

static void run_sieve(size_t n) {
	if(bits == NULL) {
		bits = malloc(n);
	}
	memset(bits, true, n);

	sieve_size = n;
	uint64_t factor = 3;
	uint64_t q = sqrt(sieve_size);

	while(factor <= q) {
		for(size_t num = factor; num < sieve_size; num += 2) {
			if(bits[num]) {
				factor = num;
				break;
			}
		}
		for(size_t num = factor * factor; num < sieve_size; num += factor * 2) {
			bits[num] = false;
		}
		factor += 2;
	}
}

static void print_results(bool show_results, double duration, uint64_t passes) {
	if(show_results) {
		printf("2, ");
	}
	size_t count = (sieve_size >= 2);
	for(size_t num = 3; num <= sieve_size; num += 2) {
		if(bits[num]) {
			if(show_results) {
				printf("%zu, ", num);
			}
			++count;
		}
	}

	if(show_results) {
		printf("\n");
	}

	printf("Passes: %" PRIu64 ", Time: %lf, Avg: %lf, Limit: %zu, Count1: %zu, Count2: %zu, Valid: %u\n",
			passes,
			duration,
			duration / passes,
			sieve_size,
			count,
			count_primes(),
			validate_results());
}

static uint64_t get_time(void) {
	struct timespec ts;
	timespec_get(&ts, TIME_UTC);
	return (ts.tv_sec * 1000000000) + ts.tv_nsec;
}

int main(void) {
	uint64_t passes = 0;
	uint64_t start = get_time();

	while(true) {
		run_sieve(1000000);
		++passes;
		if((get_time() - start) / 1000000000 >= 5) {
			print_results(false, (get_time() - start) / 1000000000, passes);
			break;
		}
	}
	return 0;
}

File: PrimePerl/primes.pl
#!/usr/bin/env perl

use strict;
use warnings;

package PrimeSieve {
    my $DICT = {
        10          => 4,
        100         => 25,
        1000        => 168,
        10000       => 1229,
        100000      => 9592,
        1000000     => 78498,
        10000000    => 664579,
        100000000   => 5761455,
        1000000000  => 50847534,
        10000000000 => 455052511,
    };

    sub new {
        my ( $class, $sieve_size ) = @_;

        my $self = {
            sieve_size => $sieve_size,
            bits       => [],
        };
        bless( $self, $class );

        return $self;
    }

    sub run_sieve {
        my $self   = shift();
        my $factor = 3;

        while ( $factor <= sqrt $self->{sieve_size} ) {
            my $num1 = $factor;
            while ( $num1 < $self->{sieve_size} ) {
                if ( !$self->{bits}->[$num1] ) {
                    $factor = $num1;
                    last;
                }

                $num1 += 2;
            }

            my $num2 = $factor * $factor;
            while ( $num2 < $self->{sieve_size} ) {
                $self->{bits}->[$num2] = 1;
                $num2 += $factor << 1;
            }

            $factor += 2;
        }
    }

    sub count_primes {
        my $self = shift();

        my $count = 1;
        for ( my $i = 3 ; $i < $self->{sieve_size} ; $i += 2 ) {
            $count++ if ( !$self->{bits}->[$i] );
        }

        return $count;
    }

    sub validate_results {
        my $self = shift();

        return $DICT->{ $self->{sieve_size} } == $self->count_primes();
    }

    sub print_results {
        my ( $self, $show_results, $duration, $passes ) = @_;

        my $count = 1;
        for ( my $num = 3 ; $num <= $self->{sieve_size} ; $num += 2 ) {
            $count++ if ( !$self->{bits}->[$num] );
        }

        printf(
"Passes: %d Time: %s Avg: %s Limit: %d Count1: %d Count2: %d Valid: %d\n",
            $passes, $duration, ( $duration / $passes ),
            $self->{sieve_size}, $count, $self->count_primes(),
            $self->validate_results()
        );
    }
}

package main {
    my $passes = 0;
    my $start  = time();

    while (1) {
        my $sieve = PrimeSieve->new(1000000);
        $sieve->run_sieve();
        $passes++;

        my $duration = time() - $start;
        if ( $duration >= 5 ) {
            $sieve->print_results( 0, $duration, $passes );
            last;
        }
    }
}
